### PR TITLE
ai code review and fix

### PR DESCRIPTION
**Pull request by pstjohn**
Misc. semi-automated fixes adding documentation and some additional tests throughout the recipes directories

## Release Notes

* **New Features**
  * Added Mixtral model support with TransformerEngine optimizations and comprehensive documentation
  * Introduced context-parallel tensor processing helpers across model implementations
  * Added token dropout support for different input formats (BSHD and THD)
  * New test utilities for checkpoint pruning and scheduler validation

* **Refactoring**
  * Updated state transformation API across convert modules for consistency
  * Simplified checkpoint and dataset utilities
  * Refactored collators to use modular helper functions

* **Documentation**
  * Expanded module docstrings describing state dict transformation system
  * Added documentation for attention input formats across models
  * Updated README references and configuration documentation

* **Tests**
  * Added Mixtral export validation tests
  * New checkpoint pruning and scheduler tests with comprehensive edge-case coverage
  * Updated test base classes for consistency across model implementations

### CODE REVIEW

File: bionemo-recipes/models/mixtral/tests/test_export.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (trvachov) [inline bionemo-recipes/models/mixtral/tests/test_export.py]: We are in 2026 now :)

File: bionemo-recipes/models/mixtral/tests/test_export.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+from export import export_hf_checkpoint
+
+
+@pytest.mark.skipif(os.getenv("CI", "false") == "true", reason="Skipping test in CI, requires Mini-Mixtral download.")
+def test_export_mixtral_checkpoint(tmp_path):
+    export_hf_checkpoint("NeuralNovel/Mini-Mixtral-v0.2", tmp_path / "checkpoint_export")
+
+    _ = AutoTokenizer.from_pretrained(tmp_path / "checkpoint_export")
+    model = AutoModelForCausalLM.from_pretrained(tmp_path / "checkpoint_export", trust_remote_code=True)
```
Comment (trvachov) [inline bionemo-recipes/models/mixtral/tests/test_export.py]: Is this "trust_remote_code" going to trigger a security notice?

Comment (trvachov) [reply bionemo-recipes/models/mixtral/tests/test_export.py]: @coderabbitai what do you think?

Comment (pstjohn) [reply bionemo-recipes/models/mixtral/tests/test_export.py]: silly coderabbit, trix are for kids

anyways, we don't have a revision here because we're loading from a local filesystem. We have to say "trust_remote_code" because we're testing whether the code we've dumped into that folder is executable, even though it's actually 'local' code in this case.

The alternative is to use the bare `NVMixtralFor...` classes, but that doesn't give us the same coverage that our "auto_map" dictionary is working correctly

File: bionemo-recipes/models/mixtral/tests/test_export.py (hunk 1 above)
Comment (pstjohn) [reply bionemo-recipes/models/mixtral/tests/test_export.py]: i'll add a comment to claude.md to tell it to not copy license headers 😄 . The script is correct, but this got copied from another file clearly.

**Comment by pstjohn** [comment]
@coderabbiai resolve

File: bionemo-recipes/models/esm2/collator.py [hunk 3]
```diff
@@ -730,6 +730,144 @@ def _pt_flatten_collate(features: list[dict[str, list[int]]], return_position_id
...
+        cp_rank: The context parallelism rank index.
+        total_slices: Total number of slices per sequence (2 * cp_world_size).
+
+    Returns:
+        The sharded tensor for the given CP rank, or None if val is None.
+    """
+    if val is None:
+        return val
+
+    seq_dim = _find_seq_dim(val, seq_len)
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/collator.py:786]: Did it just make this function simply because there was too much code? The base function that this came from was from TE

File: bionemo-recipes/models/esm2/collator.py [hunk 4]
```diff
@@ -837,110 +975,20 @@ def _split_batch_by_cp_rank(
...
-                cp_rank_slices.append(
-                    torch.arange(
-                        seq_start + (cp_rank * slice_size),
-                        seq_start + ((cp_rank + 1) * slice_size),
-                        device=val.device,
-                    )
-                )
+        # Ensure cu_seqlens_padded[-1] is a Python int, not a 0-dim tensor
+        last_elem = cu_seqlens_padded[-1]
+        seq_len_val = last_elem.item() if isinstance(last_elem, torch.Tensor) else last_elem
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/collator.py:980]: not sure what edge case this is fixing, Ideally we take this from TE anyways. https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py#L4142C5-L4142C12

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py [hunk 5]
```diff
@@ -98,13 +98,14 @@ def main(args: DictConfig) -> float | None:
...
     transformer_stack = model.esm.encoder.layers if hasattr(model.esm.encoder, "layers") else model.esm.encoder.layer
 
-    mp_policy = MixedPrecisionPolicy(
-        param_dtype=torch.bfloat16
-        if args.use_fp32_master_weights
-        else None,  # Cast params to BF16 for forward/backward
-        reduce_dtype=torch.float32 if args.use_fp32_master_weights else None,  # Gradient reductions in FP32
-        output_dtype=torch.bfloat16 if args.use_fp32_master_weights else None,  # Forward output dtype
-    )
+    if args.use_fp32_master_weights:
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py:101]: this does the same thing right? I think the left might even be cleaner it avoids the if/else loop

**Comment by jomitchellnv** [comment]
Release notes make it clear this MR covers four buckets:
	1.	Mixtral + TE support / model features (context-parallel helpers, token dropout)
	2.	Refactors (state transform API, checkpoint/dataset utils, collators)
	3.	Docs (state-dict system + input format docs + README/config)
	4.	Tests (Mixtral export validation, checkpoint/scheduler tests, base class updates)

Would you be open to splitting this into 2–4 MRs along those lines? As-is it’s difficult to review in one pass, and splitting would align better with our CONTRIBUTING guidance (“Make sure your PR does one thing/Have a clear answer to ‘What does this PR do?’")

File: bionemo-recipes/models/esm2/modeling_esm_te.py [hunk 6]
```diff
@@ -587,31 +635,10 @@ def forward(
...
-                # We need to find the number of masked tokens in each sequence in the padded batch.
-                is_masked = (input_ids == self.mask_token_id).squeeze(0)
-                n_masked_per_seq = torch.nested.nested_tensor_from_jagged(
-                    is_masked, offsets=kwargs["cu_seq_lens_q"]
-                ).sum(1)
-                mask_ratio_observed = n_masked_per_seq.float() / src_lengths
-                scale_factor = (1 - mask_ratio_train) / (1 - mask_ratio_observed)
-                reshaped_scale_factor = torch.repeat_interleave(scale_factor, src_lengths_padded, dim=0)
-                embeddings = (embeddings * reshaped_scale_factor.unsqueeze(-1)).to(embeddings.dtype)
+                embeddings = self._apply_token_dropout_bshd(embeddings, input_ids, attention_mask)
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/modeling_esm_te.py]: The old way is more readable to me

File: bionemo-recipes/models/esm2/collator.py (hunk 4 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/collator.py:980]: although I think we've diverged since then

File: bionemo-recipes/models/esm2/collator.py (hunk 3 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/collator.py:786]: https://github.com/coderabbitai what do you think?

File: bionemo-recipes/models/esm2/collator.py (hunk 4 above)
Comment (jomitchellnv) [reply bionemo-recipes/models/esm2/collator.py:980]: https://github.com/coderabbitai what edge case does this fix

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py (hunk 5 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py:101]: https://github.com/coderabbitai what do you think?

File: bionemo-recipes/models/esm2/collator.py (hunk 3 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/collator.py:786]: I think you need the "@"

@coderabbitai what do you think

File: bionemo-recipes/models/esm2/collator.py (hunk 4 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/collator.py:980]: my assumption is that this is for typing; @coderabbitai what edge case does this fix

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py (hunk 5 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py:101]: @coderabbitai what do you think?

Comment (pstjohn) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py:101]: my preference would probably just be `else: mp_policy = None`
